include/tsar/Analysis/Memory/MemorySet.h

```diff
@@ -439,6 +385,37 @@ class MemorySet {
     return IsChanged;
   }
 
+  /// \brief Clarify information about the memory locations in this set.
+  ///
+  /// \param [in] From List of pairs of memory locations, on which
+  /// clarification of information will occur. The first location of the pair is 
+  /// the original memory location, the second location is the aggregated memory 
+  /// location. If the second location does not have an exact intersection with 
+  /// any location of this set, the first location will be added to this set.
+  template<class Ty> void clarify(
+      llvm::SmallVectorImpl<std::pair<Ty, Ty>> &From) {
+    llvm::SmallVector<Ty, 4> NewLocs;
+    for (auto &Pair : From) {
+      auto &OtherLoc = Pair.second;
+      bool HasExactIntersection = false;
+      if (mLocations.find(MemoryInfo::getPtr(OtherLoc)) != mLocations.end()) {
+        for (auto &Loc : mLocations[MemoryInfo::getPtr(OtherLoc)]) {
+          auto IntOpt = MemoryInfo::intersect(Loc, OtherLoc);
+          if (IntOpt.hasValue() && MemoryInfo::getPtr(IntOpt.getValue())) {
+            HasExactIntersection = true;
+            break;
+          }
+        }
+      }
+      if (!HasExactIntersection) {
+        MemoryInfo::setNonCollapsable(Pair.first);
+        NewLocs.push_back(std::move(Pair.first));

```

Comment:
```suggestion
        NewLocs.push_back(Pair.first);
```

Перемещение не безопасно, т.к. посел него исходный объект разрушается, а т.к. это часть контейнера, то при вызове данного метода это надо помнить. При этом из прототипа этого метода такое поведение не следует, что можно привести к ошибками, которые будет сложно отловить.

Операция копирования необобщенных участков памяти - дешевая, поэтому предлагаю здесь просто использовать копирование.